File: 2016/03_SquaresWithThreeSides/triangles.ts
// ======================================================================
// Squares With Three Sides
//   Advent of Code 2016 Day 03 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           t r i a n g l e s . t s
//
// A solver for the Advent of Code 2016 Day 03 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                                  types
// ----------------------------------------------------------------------

type sidesTypes = [number, number, number];

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

// ======================================================================
//                                                              Triangles
// ======================================================================

export class Triangles {
  // Object for Squares With Three Sides
  text: string[];

  part2: boolean;

  sides: sidesTypes[];

  constructor(text: string[], part2 = false) {
    // Create a Triangles object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.sides = [];

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      this.processText();
    }
  }

  processText(): void {
    const re = /\s+/;
    this.sides = [];
    this.text.forEach((line) => {
      const sides = line.trim().split(re);
      this.sides.push([parseInt(sides[0], 10), parseInt(sides[1], 10), parseInt(sides[2], 10)]);
    });
  }

  static isTriangle(sides: sidesTypes): boolean {
    return (sides[0] + sides[1] > sides[2])
      && (sides[0] + sides[2] > sides[1])
      && (sides[1] + sides[2] > sides[0]);
  }

  numTriangles(verbose: boolean, limit: number): number {
    if (verbose) console.log(`numTriangles ${limit}`);
    let result = 0;
    this.sides.forEach((sides) => {
      if (Triangles.isTriangle(sides)) {
        if (verbose) console.log(`${sides[0]} ${sides[1]} ${sides[2]} is a trangle`);
        result += 1;
      } else if (verbose) console.log(`${sides[0]} ${sides[1]} ${sides[2]} is not a trangle`);
    });
    return result;
  }

  numTrianglesInColumns(verbose: boolean, limit: number): number {
    if (verbose) console.log(`numTriangles ${limit}`);
    let result = 0;
    for (let triples = 0; triples < this.sides.length; triples += 3) {
      for (let index = 0; index < 3; index += 1) {
        const sides: sidesTypes = [this.sides[triples][index],
          this.sides[triples + 1][index],
          this.sides[triples + 2][index]];
        if (Triangles.isTriangle(sides)) {
          if (verbose) console.log(`${sides[0]} ${sides[1]} ${sides[2]} is a trangle`);
          result += 1;
        } else if (verbose) console.log(`${sides[0]} ${sides[1]} ${sides[2]} is not a trangle`);
      }
    }
    return result;
  }

  solution(verbose = false, limit = 0): number {
    if (verbose) console.log(`solution ${limit}`);
    if (this.part2) {
      return NaN;
    }
    return NaN;
  }

  partOne(verbose = false, limit = 0): number {
    // Returns the solution for part one

    return this.numTriangles(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): number {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.numTrianglesInColumns(verbose, limit);
  }
}

// ======================================================================
// end                     t r i a n g l e s . t s                    end
// ======================================================================

File: 2016/03_SquaresWithThreeSides/triangles.py
# ======================================================================
# Squares With Three Sides
#   Advent of Code 2016 Day 03 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         t r i a n g l e s . p y
# ======================================================================
"A solver for the Advent of Code 2016 Day 03 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                                  Triangles
# ======================================================================


class Triangles(object):   # pylint: disable=R0902, R0205
    "Object for Squares With Three Sides"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        if text is None:
            self.text = []
        else:
            self.text = text

    def count_triangles(self):
        "Returns the number of triangles in the text"

        # 1. Start with nothing
        result = 0

        # 2. Loop for the rows of text
        for line in self.text:

            # 3. If the line defines a triangle, increment the count
            if Triangles.check_row(line):
                result += 1

        # 4. Return the number of triangles
        return result

    @staticmethod
    def get_numbers(line):
        "Get the numbers from the row as int[]"
        return [int(x) for x in line.replace('  ', ' ').replace('  ', ' ').strip().split(' ')]

    @staticmethod
    def check_row(line):
        "Check an input row to see if it is a trangle"

        # 1. Split out the numbers
        numbers = Triangles.get_numbers(line)

        # 2. Check it is a triangle
        return Triangles.check_sides(numbers)

    @staticmethod
    def check_sides(sides):
        "Return true if the three sides describe a triangle"

        # 1. There must be three (and only three)
        if len(sides) != 3:
            return False

        # 2. Check the sum of the sides
        if sides[0] + sides[1] <= sides[2]:
            return False
        if sides[1] + sides[2] <= sides[0]:
            return False
        if sides[0] + sides[2] <= sides[1]:
            return False

        # 3. I name the triangle
        return True

    def count_col_triangles(self):
        "Count triangles in the columns"

        # 1. Start with nothing
        result = 0
        rows = []

        # 2. Loop for the rows of text
        for line in self.text:

            # 3. Add this line to the rows
            rows.append(Triangles.get_numbers(line))

            # 4. If we have enough number, count the triangles
            if len(rows) == 3:
                if Triangles.check_sides([rows[0][0], rows[1][0], rows[2][0]]):
                    result += 1
                if Triangles.check_sides([rows[0][1], rows[1][1], rows[2][1]]):
                    result += 1
                if Triangles.check_sides([rows[0][2], rows[1][2], rows[2][2]]):
                    result += 1
                rows = []

        # 5. Return the number of triangles
        return result

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.count_triangles()

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.count_col_triangles()


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                      t r i a n g l e s . p y                     end
# ======================================================================
